### ISSUE TITLE
Support health status interface

### ISSUE THREAD

**Issue by eyalkraft** [elastic/cloudbeat#239]
**Motivation**

Once libbeat is migrated to v2 cloudbeat should support the new health status interface to report detailed health information to the user. Including requiring the user to upgrade.
This information will surfaced in the UI. which is being tracked in .

**Definition of done**

- [ ] If there's an error, cloudbeat should report it an a way viewable from Kibana
- [ ] If there's a mismatch between the integration and the cloudbeat version (which could mean missing rules/fetchers on cloudbeats side) cloudbeat should indicate the user he needs to upgrade his agent.
- [ ] make sure we understand what is the behavior once the status is not healthy - does the agent restarts cloudbeat?

**Out of scope**

What is not included in this task  

**Related tasks/epics**

- https://github.com/elastic/security-team/issues/3494
- https://github.com/elastic/package-spec/issues/165#issuecomment-1156497651
- [Design doc](https://docs.google.com/document/d/1I0PeZKIkUTaWiY2F_Dum_gSi0cpbblcftl-tz81iYak/edit)
- blocked on https://github.com/elastic/elastic-agent/issues/100

**Comment by yashtewari** [comment]
[BeatV2Manager](https://github.com/elastic/beats/blob/main/x-pack/libbeat/management/managerV2.go) with the new status/health update interfaces is not yet as mature as we'd expected. It's not being used by the "core" beats yet, as described by this issue
- https://github.com/elastic/ingest-dev/issues/1367

So it would be risky to touch it for the v8.6.0 release. I'll add more details about this in another comment.

For now, I've setup the required information/logic for version compatibility, and also started using V1 of [`Manager.UpdateStatus`](https://github.com/elastic/beats/blob/4f106cb16e58272203b411a7eed0678989919c98/libbeat/management/management.go#L66) so that it may be easier to update to V2 going forward.

- https://github.com/elastic/integrations/pull/4695
- https://github.com/elastic/cloudbeat/pull/538

The current behavior is the same as any other error in cloudbeat: the error is logged and cloudbeat is restarted by the Agent. But this is only tentative and we should make a decision on it.

**Comment by yashtewari** [comment]
I was able to send a `Degraded` signal using the V1 interface and make the Agent status on Kibana change to "Unhealthy".

![Screenshot 2022-11-23 at 5 04 23 PM](https://user-images.githubusercontent.com/9020112/203579722-919b2dba-2fc3-4413-8055-41fb1a3ff723.png)

We have two options now:
#### 1. Fail fast: cloudbeat aborts when it is running an incompatible version
Pro: It makes doing upgrades to compatible versions "compulsory" in a way, so the user will immediately know that this action needs to be taken.
Con: If we change direction in the future on how we want to deal with version compatibility, we will have to still keep accounting for the current versions that abort on having a incompatible version.

#### 2. Mark cloudbeat as "degraded" when running incompatible versions but keep going
Pro: Doesn't disrupt flow and is more "forward compatible".
Con: Currently the status change to "degraded" doesn't reveal too much information on Kibana to the user. It's not clear from the UI which integration specifically makes the Agent "unhealthy", and so the it's not easy to look at the right logs, find the problem, and rectify it.

**Comment by eyalkraft** [comment]
2 seems closer to what we ant to end up with when the v2 health status reports work as expected

**Comment by oren-zohar** [comment]
@eyalkraft @yashtewari agreed, let's go with the second option

**Comment by yashtewari** [comment]
Sounds good, PRs have been updated.

**Comment by olegsu** [comment]
I also think the second approach is better but it might confuse the users.
For example (theoretically):
1. Running `8.5` stacks (`kibana` and `cloudbeat`) and upgrading the stack to `8.6`.
2. As a result, the integration will also be bumped due to auto-grade policy.
3. The new integration specifies `cloudbeat: ">= 8.6.0 <= 8.7.0"`.
4. `cloudbeat` reports a` degraded` status when the only thing that was changed was the stack (from users point of view). The only way to "fix" this is to upgrade all the agents in that policy, which will take time and work.

**Comment by eyalkraft** [comment]
@olegsu In case the specific benchmark wasn't effected, the integration shouldn't require newer version of cloudbeat.

I'm not sure where in the integration you decided to include 

> 3. The new integration specifies `cloudbeat: ">= 8.6.0 <= 8.7.0"`.

But ideally it belongs to a policy template and then for KSPM (which didn't change) you can specify
`cloudbeat: ">= 8.5.0 <= 8.7.0"`
And for CSPM
`cloudbeat: ">= 8.6.0 <= 8.7.0"`

**Comment by olegsu** [comment]
> I'm not sure where in the integration you decided to include

I was referring the template, where the version will be [added](https://github.com/elastic/integrations/pull/4695).


> @olegsu In case the specific benchmark wasn't effected, the integration shouldn't require newer version of cloudbeat.

You said it even better, when the benchmark was updated. I think the issue is that it will be updated regardless users knowledge. The only thing that was done from his point of view is to upgrade the Elastic stack, and for some reason now KSPM integration is degraded.

**Comment by eyalkraft** [comment]
@olegsu but specific policy templates (KSPM) including their cloudbeat version requirements can remain intact, even if we upgrade the integration package as a whole

**Comment by olegsu** [comment]
Update after sync with @eyalkraft 
Instead of check versions compitability, `cloudbeat` will check that all the benchmark types that are regocnized (part of hard-coded list).
When unknown type will be added, `cloudbeat` will report `degraded` status.
This will answer on the case [above](https://github.com/elastic/cloudbeat/issues/239#issuecomment-1346837769).
* Closing https://github.com/elastic/integrations/pull/4695

**Comment by olegsu** [comment]
This is not trivial to test on BC as it either requires changing the integration/cloudbeat to custom version.

> If there's an error, cloudbeat should report it in a way viewable from Kibana

This was not addressed in a general manner, we [report](https://github.com/elastic/cloudbeat/blob/main/launcher/launcher.go#L249) errors on re-configuration issues only.

We have two more categories of issues that we might want to address
1. Initialization issues: such as non-valid credentials. Cloudbeat currently crashes with a message in the log.
2. Runtime issues: with the same example, the credentials might get invalid/expired after some period of time. I am not sure what will be the outcome of this.

**Comment by cmacknz** [comment]
> elastic-agent status shows that one component is in degradad status but every running beat is healthy (this might be a bug @cmacknz)

Run `elastic-agent status --output=json` to see the entire status report. We hide this by default but we shouldn't https://github.com/elastic/elastic-agent/issues/2107.
